File: src/bmsx/input/keyboardinput.ts
import { $ } from '../core/game';
import { getPressedState, Input, makeButtonState, resetObject } from './input';
import type { ButtonState, InputHandler, KeyboardButtonId, KeyOrButtonId2ButtonState, VibrationParams } from './inputtypes';


/**
 * Represents a keyboard input handler that implements the IInputHandler interface.
 *
 * This class manages the state of keyboard keys, allowing for key press detection,
 * consumption of key events, and resetting of input states. It listens for keydown
 * and keyup events to update the state of keys accordingly.
 *
 * @implements {InputHandler}
 */
export class KeyboardInput implements InputHandler {
	/**
	 * The state of each keyboard key.
	 */
	public keyStates: KeyOrButtonId2ButtonState = {};

	public gamepadButtonStates: KeyOrButtonId2ButtonState = {};

	public get supportsVibrationEffect(): boolean {
		return false; // Keyboard does not support vibration effects
	}

	public applyVibrationEffect(_params: VibrationParams): void {
		// No vibration effect for keyboard
	}

	/**
	 * The index of the input device, which defaults to 0 (the main player).
	 */
	public readonly gamepadIndex = 0;

	private nextPressId = 1;

	constructor(public readonly deviceId: string = 'keyboard:0') {
		this.keyStates = {};
		this.gamepadButtonStates = {};
		this.reset();
	}

	/**
	 * Resets the state of all input keys and gamepad buttons.
	 * @param except An optional array of keys or buttons to exclude from the reset.
	 */
	public reset(except?: string[]): void {
		if (!except) {
			this.keyStates = {};
			this.gamepadButtonStates = {};
		}
		else {
			resetObject(this.keyStates, except);
			resetObject(this.gamepadButtonStates, except);
		}
	}

	/**
	 * Marks the specified key as consumed, preventing further processing of its state.
	 *
	 * @param key - The identifier of the key to be consumed.
	 * @returns void
	 */
	public consumeButton(key: string, options?: { sticky?: boolean }): void {
		const sticky = options?.sticky ?? true;
		const state = this.gamepadButtonStates[key] ?? (this.gamepadButtonStates[key] = makeButtonState());
		state.consumed = true;
		state.stickyConsumed = sticky ? true : state.stickyConsumed ?? false;
		// Use the constant to map keyboard keys to gamepad buttons
		const keyMappedToCorrespondingGamepadButtonId = Input.KEYBOARDKEY2GAMEPADBUTTON[key as keyof typeof Input.KEYBOARDKEY2GAMEPADBUTTON];
		if (keyMappedToCorrespondingGamepadButtonId) {
			const mappedState = this.gamepadButtonStates[keyMappedToCorrespondingGamepadButtonId] ?? (this.gamepadButtonStates[keyMappedToCorrespondingGamepadButtonId] = makeButtonState());
			mappedState.consumed = true;
			mappedState.stickyConsumed = sticky ? true : mappedState.stickyConsumed ?? false;
		}
	}

	/**
	 * Retrieves the current state of a specified button.
	 *
	 * @param key - The identifier for the button whose state is to be retrieved.
	 * @returns The current state of the button as a ButtonState object.
	 *          If the provided key is null, a default ButtonState is returned.
	 */
	public getButtonState(key: string): ButtonState {
		if (key === null) return makeButtonState();
		return getPressedState(this.gamepadButtonStates, key);
		// const convertedKey = Input.KEYBOARDKEY2GAMEPADBUTTON[key] ? Input.KEYBOARDKEY2GAMEPADBUTTON[key] : key;
		// return getPressedState(this.gamepadButtonStates, convertedKey);
	}

	/**
	 * Polls the input from the keyboard.
	 * This function should be called once per frame to ensure that keyboard input is up-to-date.
	 * It updates the state of each key based on the current keydown and keyup events.
	 * @returns void
	 */
	pollInput(): void {
		const now = $.platform.clock.now();
		// Update existing keys in place, create states on demand
		Object.keys(this.keyStates).forEach(buttonId => {
			const prev = this.gamepadButtonStates[buttonId] ?? makeButtonState();
			const isDown = this.keyStates[buttonId].pressed === true;
			const wasDown = prev.pressed === true;

			let pressId = prev.pressId ?? null;
			if (isDown && !pressId) {
				pressId = this.nextPressId++;
			}
			const pressedAt = wasDown ? (prev.pressedAtMs ?? prev.timestamp ?? now) : now;

			let state: ButtonState;
			if (isDown) {
				const stickyConsumed = prev.stickyConsumed ?? false;
				state = {
					...prev,
					pressed: true,
					justpressed: !wasDown,
					justreleased: false,
					waspressed: true,
					wasreleased: prev.wasreleased,
					presstime: Math.max(0, now - pressedAt),
					pressedAtMs: pressedAt,
					releasedAtMs: null,
					timestamp: wasDown ? (prev.timestamp ?? pressedAt) : now,
					pressId,
					value: 1,
					consumed: stickyConsumed,
					stickyConsumed,
				};
			} else {
				state = {
					...prev,
					pressed: false,
					justpressed: false,
					justreleased: wasDown,
					waspressed: prev.waspressed || wasDown,
					wasreleased: prev.wasreleased || wasDown,
					presstime: null,
					pressedAtMs: null,
					releasedAtMs: wasDown ? now : prev.releasedAtMs ?? null,
					timestamp: wasDown ? now : prev.timestamp ?? now,
					pressId: wasDown ? (prev.pressId ?? pressId) : null,
					value: 0,
					consumed: false,
					stickyConsumed: false,
				};
			}

			this.gamepadButtonStates[buttonId] = state;

			const mapped = Input.KEYBOARDKEY2GAMEPADBUTTON[buttonId as keyof typeof Input.KEYBOARDKEY2GAMEPADBUTTON];
			if (mapped) {
				const dst = this.gamepadButtonStates[mapped] ?? makeButtonState();
				dst.pressed = state.pressed;
				dst.justpressed = state.justpressed;
				dst.justreleased = state.justreleased;
				dst.waspressed = state.waspressed;
				dst.wasreleased = state.wasreleased;
				dst.consumed = state.consumed;
				dst.presstime = state.presstime;
				dst.timestamp = state.timestamp;
				dst.pressedAtMs = state.pressedAtMs;
				dst.releasedAtMs = state.releasedAtMs;
				dst.pressId = state.pressId;
				dst.value = state.value;
				dst.value2d = state.value2d ?? null;
				dst.stickyConsumed = state.stickyConsumed;
				this.gamepadButtonStates[mapped] = dst;
			}
		});
	}

	public ingestButton(_code: string, _state: ButtonState): void { }

	/**
	 * Sets the key state to true when a key is pressed.
	 * @param key_code - The button ID or string representing the key.
	 */
	keydown(key_code: KeyboardButtonId | string): void {
		if (!this.keyStates[key_code]) {
			this.keyStates[key_code] = makeButtonState({ pressed: true, justpressed: true, presstime: 0, timestamp: $.platform.clock.now() });
		} else {
			this.keyStates[key_code].pressed = true;
		}
	}

	/**
	 * Handles the keyup event for a given key.
	 * @param key_code - The key identifier or name.
	 */
	keyup(key_code: KeyboardButtonId | string): void {
		if (!this.keyStates[key_code]) return;
		this.keyStates[key_code] = makeButtonState();
	}

	/**
	 * Handles the blur event of the input element by resetting the input state.
	 * @param _e - The blur event object.
	 */
	blur(_e: FocusEvent): void {
		// this.preventInput = true; // Prevent input when the window loses focus
		this.reset();
	}

	/**
	 * Handles the focus event for the input element by resetting the input state.
	 * Resets the input state.
	 * @param _e - The focus event object.
	 */
	focus(_e: FocusEvent): void {
		this.reset();
		// this.preventInput = false; // Allow input when the window regains focus
	}

	dispose(): void {
		this.reset();
		this.keyStates = {};
		this.gamepadButtonStates = {};
	}
}

File: src/bmsx_cpp/input/keyboardinput.cpp
/*
 * keyboardinput.cpp - Keyboard input handler implementation
 *
 * Mirrors TypeScript input/keyboardinput.ts
 */

#include "keyboardinput.h"
#include "input.h"
#include "../core/engine_core.h"
#include <algorithm>

namespace bmsx {

/* ============================================================================
 * Constructor / Destructor
 * ============================================================================ */

KeyboardInput::KeyboardInput(const std::string& deviceId)
	: m_deviceId(deviceId)
{
	reset();
}

KeyboardInput::~KeyboardInput() {
	dispose();
}

/* ============================================================================
 * InputHandler interface
 * ============================================================================ */

void KeyboardInput::pollInput() {
	m_currentTimeMs = EngineCore::instance().clock()->now();
	
	// Update gamepad button states from key states
	for (auto& [keyCode, keyState] : m_keyStates) {
		auto& prev = m_gamepadButtonStates[keyCode];
		bool isDown = keyState.pressed;
		bool wasDown = prev.pressed;
		
		i32 pressId = keyState.pressId.value_or(prev.pressId.value_or(0));
		if (isDown && !pressId) {
			pressId = m_nextPressId++;
		}
		
		f64 pressedAt = wasDown 
			? prev.pressedAtMs.value_or(prev.timestamp.value_or(m_currentTimeMs))
			: m_currentTimeMs;
		
		ButtonState state;
		if (isDown) {
			bool stickyConsumed = prev.consumed;
			state.pressed = true;
			state.justpressed = !wasDown;
			state.justreleased = false;
			state.waspressed = true;
			state.wasreleased = prev.wasreleased;
			state.presstime = std::max(0.0, m_currentTimeMs - pressedAt);
			state.pressedAtMs = pressedAt;
			state.releasedAtMs = std::nullopt;
			state.timestamp = wasDown ? prev.timestamp.value_or(pressedAt) : m_currentTimeMs;
			state.pressId = pressId;
			state.value = 1.0f;
			state.consumed = stickyConsumed;
		} else {
			state.pressed = false;
			state.justpressed = false;
			state.justreleased = wasDown;
			state.waspressed = prev.waspressed || wasDown;
			state.wasreleased = prev.wasreleased || wasDown;
			state.presstime = std::nullopt;
			state.pressedAtMs = std::nullopt;
			state.releasedAtMs = wasDown ? m_currentTimeMs : prev.releasedAtMs;
			state.timestamp = wasDown ? m_currentTimeMs : prev.timestamp.value_or(m_currentTimeMs);
			state.pressId = wasDown ? prev.pressId : std::nullopt;
			state.value = 0.0f;
			state.consumed = false;
		}
		
		m_gamepadButtonStates[keyCode] = state;
		
		// Map to gamepad button if applicable
		const auto& mapping = Input::KEYBOARD_TO_GAMEPAD();
		auto it = mapping.find(keyCode);
		if (it != mapping.end()) {
			const std::string& mappedButton = it->second;
			auto& dst = m_gamepadButtonStates[mappedButton];
			
			// Only update if not already pressed by another key
			if (!dst.pressed || state.pressed) {
				dst.pressed = state.pressed;
				dst.justpressed = state.justpressed;
				dst.justreleased = state.justreleased;
				dst.waspressed = state.waspressed;
				dst.wasreleased = state.wasreleased;
				dst.consumed = state.consumed;
				dst.presstime = state.presstime;
				dst.timestamp = state.timestamp;
				dst.pressedAtMs = state.pressedAtMs;
				dst.releasedAtMs = state.releasedAtMs;
				dst.pressId = state.pressId;
				dst.value = state.value;
				dst.value2d = state.value2d;
			}
		}
	}
}

ButtonState KeyboardInput::getButtonState(const ButtonId& button) {
	auto it = m_gamepadButtonStates.find(button);
	if (it == m_gamepadButtonStates.end()) {
		return ButtonState{};
	}
	return it->second;
}

void KeyboardInput::consumeButton(const ButtonId& button) {
	// Consume the button state
	auto it = m_gamepadButtonStates.find(button);
	if (it != m_gamepadButtonStates.end()) {
		it->second.consumed = true;
	}
	
	// Also consume any keyboard key that maps to this gamepad button
	const auto& mapping = Input::KEYBOARD_TO_GAMEPAD();
	for (const auto& [keyCode, mappedButton] : mapping) {
		if (mappedButton == button) {
			auto keyIt = m_gamepadButtonStates.find(keyCode);
			if (keyIt != m_gamepadButtonStates.end()) {
				keyIt->second.consumed = true;
			}
		}
	}
}

void KeyboardInput::reset(const std::vector<std::string>* except) {
	if (!except) {
		m_keyStates.clear();
		m_gamepadButtonStates.clear();
	} else {
		resetObjectMap(m_keyStates, except);
		resetObjectMap(m_gamepadButtonStates, except);
	}
}

void KeyboardInput::dispose() {
	reset();
}

/* ============================================================================
 * Key events
 * ============================================================================ */

void KeyboardInput::keydown(const std::string& keyCode, i32 pressId, f64 timestamp) {
	auto& state = m_keyStates[keyCode];
	state.pressed = true;
	state.justpressed = true;
	state.presstime = 0.0;
	state.timestamp = timestamp;
	state.pressedAtMs = timestamp;
	state.pressId = pressId;
}

void KeyboardInput::keyup(const std::string& keyCode, i32 /*pressId*/, f64 /*timestamp*/) {
	m_keyStates[keyCode] = ButtonState{};
}

void KeyboardInput::blur() {
	reset();
}

void KeyboardInput::focus() {
	reset();
}

} // namespace bmsx
